File: f1tenth_waypoint/src/waypoint_generator/record_odom_waypoints.cpp
#include "f1tenth_waypoint/record_odom_waypoints.hpp"
#include <cmath> 

OdomWaypoint::OdomWaypoint(std::shared_ptr<rclcpp::Node> rclcpp_node) : UtilsWaypoint(rclcpp_node) , 
                                                                        node_(rclcpp_node)
{


    try
    {
        if (!node_->has_parameter("odom_min_distance"))
        {
            node_->declare_parameter<double>("odom_min_distance", 0.05);
        }
        min_distance = this->node_->get_parameter("odom_min_distance").as_double();
        RCLCPP_INFO(this->node_->get_logger(), "min distance parameter set to: %.2f", min_distance);

    }
    catch(const rclcpp::exceptions::ParameterNotDeclaredException &e) 
    {
        RCLCPP_WARN(this->node_ -> get_logger(), "Could not get parameters ... setting variables to default");
        RCLCPP_WARN(this->node_ -> get_logger(), "Error: %s", e.what());
    }
    catch(const std::exception &e)
    {
        RCLCPP_WARN(this->node_ ->get_logger(), "An error occured: %s", e.what());
        
    }
    
    this->initialize(rclcpp_node);
    
    
}

void OdomWaypoint::initialize(std::shared_ptr<rclcpp::Node> rclcpp_node_)
{
    csv_file_ = std::make_shared<CSVFile>(rclcpp_node_);

   // csv_file_ = std::make_shared<CSVFile>(this->shared_from_this());
}


void OdomWaypoint::odom_callback(const nav_msgs::msg::Odometry::ConstSharedPtr odom_msgs,
                                const rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr& waypoint_marker_pub_)
{
    double diff = sqrt(pow((odom_msgs->pose.pose.position.x - x_old), 2) + 
                        pow((odom_msgs->pose.pose.position.y - y_old), 2));

    if(diff > min_distance)
    {
        RCLCPP_INFO(this->node_->get_logger(), "Received odometry message: x = %.2f, y = %.2f",
                     odom_msgs->pose.pose.position.x, odom_msgs->pose.pose.position.y);

         //auto new_point_shared = std::make_shared<geometry_msgs::msg::Point>(new_point);

        this->write_waypoint_to_csv(odom_msgs);
        this->update_waypoint_marker_callback(odom_msgs); 
        this->public_waypoint_marker(waypoint_marker_pub_);
        

        // update old positions
        this->x_old = odom_msgs->pose.pose.position.x;
        this->y_old = odom_msgs->pose.pose.position.y;
        

    }
}

void OdomWaypoint::public_waypoint_marker(const rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr& waypoint_marker_pub_)
{
    const auto& waypoint_marker = this->get_waypoint_marker();
    waypoint_marker_pub_->publish(waypoint_marker);
}



OdomWaypoint::~OdomWaypoint()
{

}

File: f1tenth_waypoint/f1tenth_waypoint_generator/waypoint_generation_options/record_odom_waypoints.py

import math
from rclpy.node import Node, Publisher
from nav_msgs.msg import Odometry
from  f1tenth_waypoint_generator.utils.waypoint_utils import UtilsWaypoint



class OdomWaypoint(UtilsWaypoint):
    def __init__(self, rclpy_node: Node) -> None:
        super().__init__(rclpy_node)

        self.rclpy_node = rclpy_node


        if not self.rclpy_node.has_parameter('odom_min_distance'):
            rclpy_node.declare_parameter(name='odom_min_distance', value=0.05)

        try:
            self.min_distance = rclpy_node.get_parameter('odom_min_distance').get_parameter_value().double_value

        except Exception as e:
            self.rclpy_node.get_logger().warn('Could not get parameters ...setting variables to default')
            self.rclpy_node.get_logger().warn(f'Error: "{e}"')
            self.min_distance = 0.05
        
        self.x_old = 0; 
        self.y_old = 0
    

    def odom_callback(self, odom_msg:Odometry, waypoint_marker_pub_:Publisher)->None:
        
        # Calculate the difference using proper distance formula
        diff = math.sqrt(
                            pow((odom_msg.pose.pose.position.x - self.x_old), 2) + 
                            pow((odom_msg.pose.pose.position.y - self.y_old), 2))
        
        # check if the car has moved more than the minimum distance 
        if diff > self.min_distance:
            # Up date car position
            x_car = odom_msg.pose.pose.position.x
            y_car = odom_msg.pose.pose.position.y 
            self.rclpy_node.get_logger().info(f'Car moved: x = {x_car}, y = {y_car}, distance = {diff:.2f} meters')
            self.process_waypoint(odom_msg)
            self.update_waypoint_marker(odom_msg)
            self.publish_waypoint_marker(waypoint_marker_pub_)

            # Update old position values
            self.x_old = x_car
            self.y_old = y_car
    
    def publish_waypoint_marker(self,  waypoint_marker_pub_:Publisher):
        """Publish the updated marker to the topic"""
        waypoint_marker = self.get_waypoint_marker()
        waypoint_marker_pub_.publish(waypoint_marker)






